Rubberduck.Parsing/VBA/RubberduckParserState.cs

```diff
@@ -209,49 +209,23 @@ public void OnStatusMessageUpdate(string message)
             }
         }
 
-        public void AddProject(string projectId)
+        /// <summary>
+        /// Refreshes our list of cached projects.
+        /// Be sure to reparse after calling this in case there
+        /// were projects with duplicate ID's to clear the old
+        /// declarations referencing the project by the old ID.
+        /// </summary>
+        public void RefreshProjects()
         {
-            var projects = new List<VBProject>();
-            foreach (VBProject p in _vbe.VBProjects)
+            _projects.Clear();
+            foreach (VBProject project in _vbe.VBProjects)
             {
-                if (p.HelpFile != null && _projects.Keys.Contains(p.HelpFile))
+                if (string.IsNullOrEmpty(project.HelpFile) || _projects.Keys.Contains(project.HelpFile))
                 {
-                    continue;
+                    project.AssignProjectId();
                 }
 
-                projects.Add(p);
-            }
-
-            if (projects.Count == 0)
-            {
-                Logger.Debug("Project was not found and will not be added to parser state.");
-                return;
-            }
-
-            foreach (var project in projects)
-            {
-                if (project.Protection == vbext_ProjectProtection.vbext_pp_locked)
-                {
-                    // adding protected project to parser state is asking for COMExceptions..
-                    Logger.Debug("Project is protected and will not be added to parser state.");
-                    return;
-                }
-
-                if (string.IsNullOrEmpty(project.HelpFile))
-                {
-                    // assigns the help file and returns the value to reduce COM calls
-                    projectId = project.AssignProjectId();
-                }
-
-                if (!_projects.ContainsKey(projectId))
-                {
-                    _projects.Add(projectId, project);
-                }
-
-                foreach (VBComponent component in project.VBComponents)
-                {
-                    _moduleStates.TryAdd(new QualifiedModuleName(component), new ModuleState(ParserState.Pending));
-                }
+                _projects.Add(project.HelpFile, project);
```

Comment:
lol, an extension method would do it. as for extension properties... we're not on C# 6, so even if you implement it and it ships with C# 8... :wink: